p2p/protocol/holepunch/coordination.go

```diff
@@ -49,11 +50,53 @@ func NewHolePunchService(h host.Host, ids *identify.IDService) (*HolePunchServic
 	ctx, cancel := context.WithCancel(context.Background())
 	hs := &HolePunchService{ctx: ctx, ctxCancel: cancel, host: h, ids: ids}
 
+	sub, err := h.EventBus().Subscribe(new(event.EvtNATDeviceTypeChanged))
+	if err != nil {
+		return nil, err
+	}
+
 	h.SetStreamHandler(protocol, hs.handleNewStream)
 	h.Network().Notify((*netNotifiee)(hs))
+
+	hs.refCount.Add(1)
+	go hs.loop(sub)
+
 	return hs, nil
 }
 
+func (hs *HolePunchService) loop(sub event.Subscription) {
+	defer hs.refCount.Done()
+	defer sub.Close()
+
+	for {
+		select {
+		case _, ok := <-sub.Out():
+			if !ok {
+				return
+			}
+
+			if hs.peerSupportsHolePunching(hs.host.ID(), hs.host.Addrs()) {
+				hs.host.SetStreamHandler(protocol, hs.handleNewStream)
+			} else {
+				hs.host.RemoveStreamHandler(protocol)
```

Comment:
wait, this is a BUG -- we lose the ability to dial back connect to a public node that has dialed us.